draft-wang-ppm-dap-taskprov.md

```diff
@@ -56,76 +60,330 @@ collector). Thus it is up to each deployment of DAP to securely implement its
+the upload protocol and alert the client with error "unrecognizedMessage".
+
+> reason for aborting: if leader ignores it, then there is no way to ensure
+> clients that the extension is used and task parameters are respected in
+> aggregators
+
+If leader supports `task-prov` extension, it should first check if the task ID
+already exists, if so leader continues to the rest of upload flow as usual.
+Note if the existing tasks's configuration is different from the one in report
+extension, HPKE decryption will fail due to mismatched AAD.
+
+If the task ID has not been seen before, leader should read and decode report's
+`extension_data` with the `TaskConfig` schema. If the decoding failed, it MUST
+abort the upload protocol and alert the client with error
+"unrecognizedMessage".
+
+If the decode succeeds, leader should create a new task using the task ID from
+the decoded report, and save task configuration with the newly created task. In
+particular, leader should deserialize `vdaf_config` corresponding to `vdaf_type`,
+and pass the relevant parameters to the VDAF initializer. At this point, the
+task provision step has completed, and leader should continue to the rest of
+upload flow.
+
+Leader MAY return error to the client if task creation failed.
+
+> OPEN ISSUE: would returning error reveal the position of the client in the
+> batch of reports received? if so is this a security or privacy threat?
+
+## Change to leader aggregate sub-protocol
+
+There is no change to the aggregate sub-protocol on leader side.
+
+## Change to collect sub-protocol
+
+During collect init, leader must perform batch validation (see batch-validation
+in {{?DAP=I-D.draft-ietf-ppm-dap-01}}) to the `CollectReq`. This requires leader
+to know the parameters associated with the DAP task. However, upon receipt of
+`CollectReq`, leader may not know the parameters if the first upload has not
+arrived yet. In this case the leader MAY respond with HTTP status code 404
+Not Found and an error of type `unrecognizedTask`. The response MAY include a
+Retry-After header field to suggest a pulling interval to the collector.
+
+> Alternatively the leader can just response with 303 and continue waiting.
 
 # Helper Behavior
 
-> TODO(wangshan) Say what the Helper does when it receives a report share with
-> this extension. (Same questions as above.)
+Helper should have received out-of-band from leader any information that is not
+always tied to a particular task, for e.g. `vdaf_verify_key`. Similar to
+leader, helper may not know the task ID of the current task before receiving
+the first `AggregateInitializeReq` message from the leader.
+
+## Change to helper aggregate sub-protocol
 
+Upon receipt of a `AggregateInitializeReq`, helper reads the extension
+codepoint in `extension_type`. If helper does not support this extension, it
+SHOULD abort the aggregate protocol and alert the leader with error
+"unrecognizedMessage".
```

Comment:
Again: Let's assume that the Helper MUST ignore unrecognized extensions. See https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/334

Reply:
what does "ignore" mean in this case? it should carry on preparing the shares as if task ID is already associated with an existing task? The current core protocol doesn't specify what happens in the aggregate-flow if the task ID is not recognized. Maybe we should specify that behaviour in the core protocol first, then we can simply see "ignore" in this extension.

Reply:
If the Helper doesn't recognize the task ID, it's supposed to abort with "unrecognizedTask". Are you sure this isn't already spelled out?